### PR TITLE
[JBPM-9880] logging async node enter

### PR DESCRIPTION
**Pull request by elguardian**
Jira: https://issues.redhat.com/browse/JBPM-9880
depends On https://github.com/kiegroup/droolsjbpm-knowledge/pull/547

### CODE REVIEW

**Comment by elguardian** [comment]
Jenkins retest this please

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/mysql5/jbpm-7.59-to-7.60.sql [hunk 1]
```diff
@@ -3,3 +3,4 @@ alter table BAMTaskSummary drop column end_date;
 alter table TaskEvent drop column end_date;
 alter table NodeInstanceLog drop column end_date;
 alter table VariableInstanceLog drop column end_date;
+alter table NodeInstanceLog add column observation varchar(255);
```
Comment (afalhambra) [inline jbpm-db-scripts/src/main/resources/db/upgrade-scripts/mysql5/jbpm-7.59-to-7.60.sql:6]: "observation" field in the create script for mysql5 is missing

**Review by afalhambra — CHANGES_REQUESTED** [review]
Looks good to me, 
a few comments in the test case. 
Also please correct mysql DDL script cause it is missing "observation" new field.

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/oracle/rhpam-7.12-to-7.13.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+alter table NodeInstanceLog add column observation varchar2(255 char);
```
Comment (afalhambra) [inline jbpm-db-scripts/src/main/resources/db/upgrade-scripts/oracle/rhpam-7.12-to-7.13.sql:1]: same

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/oracle/jbpm-7.59-to-7.60.sql [hunk 3]
```diff
@@ -3,3 +3,4 @@ alter table BAMTaskSummary drop column end_date;
 alter table TaskEvent drop column end_date;
 alter table NodeInstanceLog drop column end_date;
 alter table VariableInstanceLog drop column end_date;
+alter table NodeInstanceLog add column observation varchar2(255 char);
```
Comment (afalhambra) [inline jbpm-db-scripts/src/main/resources/db/upgrade-scripts/oracle/jbpm-7.59-to-7.60.sql:6]: `column` is an invalid identifier in oracle - please remove
```suggestion
alter table NodeInstanceLog add observation varchar2(255 char);
```

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver/jbpm-7.59-to-7.60.sql [hunk 4]
```diff
@@ -3,3 +3,4 @@ alter table BAMTaskSummary drop column end_date;
 alter table TaskEvent drop column end_date;
 alter table NodeInstanceLog drop column end_date;
 alter table VariableInstanceLog drop column end_date;
+alter table NodeInstanceLog add column observation varchar(255);
```
Comment (afalhambra) [inline jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver/jbpm-7.59-to-7.60.sql:6]: `column` is an invalid identifier in mssql - please remove
```suggestion
alter table NodeInstanceLog add observation varchar(255);
```

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver/rhpam-7.12-to-7.13.sql [hunk 5]
```diff
@@ -0,0 +1,3 @@
+alter table NodeInstanceLog add column observation varchar(255);
```
Comment (afalhambra) [inline jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver/rhpam-7.12-to-7.13.sql:1]: same

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver2008/jbpm-7.59-to-7.60.sql [hunk 6]
```diff
@@ -3,3 +3,4 @@ alter table BAMTaskSummary drop column end_date;
 alter table TaskEvent drop column end_date;
 alter table NodeInstanceLog drop column end_date;
 alter table VariableInstanceLog drop column end_date;
+alter table NodeInstanceLog add column observation varchar(255);
```
Comment (afalhambra) [inline jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver2008/jbpm-7.59-to-7.60.sql:6]: same

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver2008/rhpam-7.12-to-7.13.sql [hunk 7]
```diff
@@ -0,0 +1,3 @@
+alter table NodeInstanceLog add column observation varchar(255);
```
Comment (afalhambra) [inline jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver2008/rhpam-7.12-to-7.13.sql:1]: same

**Review by afalhambra — CHANGES_REQUESTED** [review]
besides previous comment, there are some other issues found for the upgrade scripts in oracle and mssql server

File: jbpm-audit/src/main/java/org/jbpm/process/audit/jms/AsyncAuditLogProducer.java [hunk 8]
```diff
@@ -123,6 +124,13 @@ public void beforeNodeTriggered(ProcessNodeTriggeredEvent event) {
         ((NodeInstanceImpl) event.getNodeInstance()).getMetaData().put("NodeInstanceLog", log);
     }
 
+    @Override
+    public void onAsyncNodeEventScheduled(ProcessAsyncNodeScheduledEvent event) {
+        NodeInstanceLog log = (NodeInstanceLog) builder.buildEvent(event);
+        sendMessage(log, ON_ASYNC_NODE_EVENT_TYPE, 7);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-audit/src/main/java/org/jbpm/process/audit/jms/AsyncAuditLogProducer.java:130]: BEFORE_NODE_ENTER_EVENT_TYPE has priority 8 so this one should be get before that one.

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/AsyncEventNodeInstance.java [hunk 9]
```diff
@@ -17,9 +17,14 @@
 package org.jbpm.workflow.instance.node;
 
 import java.io.Serializable;
+import java.util.Collections;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/AsyncEventNodeInstance.java:20]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/async/AsyncThreadIsolationTest.java [hunk 10]
```diff
@@ -104,6 +109,14 @@ public void testCorrectProcessStateAfterException() {
         pi = ksession.getProcessInstance(pi.getId());
         Assertions.assertThat(pi).isNotNull();
         Assertions.assertThat(pi.getState()).isEqualTo(ProcessInstance.STATE_ACTIVE);
+
+        EntityManager em = getEmf().createEntityManager();
+        List<NodeInstanceLog> logs = em.createQuery("SELECT o FROM NodeInstanceLog o WHERE o.type = :type", NodeInstanceLog.class)
+                .setParameter("type", NodeInstanceLog.TYPE_ASYNC_ENTER)
+                .getResultList();
+        em.close();
+        assertTrue(logs.size() > 0);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/async/AsyncThreadIsolationTest.java:118]: done

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/mysql5/jbpm-7.59-to-7.60.sql (hunk 1 above)
Comment (elguardian) [reply jbpm-db-scripts/src/main/resources/db/upgrade-scripts/mysql5/jbpm-7.59-to-7.60.sql:6]: added

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/oracle/jbpm-7.59-to-7.60.sql (hunk 3 above)
Comment (elguardian) [reply jbpm-db-scripts/src/main/resources/db/upgrade-scripts/oracle/jbpm-7.59-to-7.60.sql:6]: added

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver/jbpm-7.59-to-7.60.sql (hunk 4 above)
Comment (elguardian) [reply jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver/jbpm-7.59-to-7.60.sql:6]: ok

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver/rhpam-7.12-to-7.13.sql (hunk 5 above)
Comment (elguardian) [reply jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver/rhpam-7.12-to-7.13.sql:1]: ok

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver2008/jbpm-7.59-to-7.60.sql (hunk 6 above)
Comment (elguardian) [reply jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver2008/jbpm-7.59-to-7.60.sql:6]: ok

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver2008/rhpam-7.12-to-7.13.sql (hunk 7 above)
Comment (elguardian) [reply jbpm-db-scripts/src/main/resources/db/upgrade-scripts/sqlserver2008/rhpam-7.12-to-7.13.sql:1]: ok

File: jbpm-db-scripts/src/main/resources/db/upgrade-scripts/oracle/rhpam-7.12-to-7.13.sql (hunk 2 above)
Comment (elguardian) [reply jbpm-db-scripts/src/main/resources/db/upgrade-scripts/oracle/rhpam-7.12-to-7.13.sql:1]: ok

**Review by afalhambra — APPROVED** [review]
Outstanding work

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

**Pull request closed without merging**